tracer/src/Datadog.Trace/Iast/Ranges.cs

```diff
@@ -206,22 +207,42 @@ internal static Range[] CopyWithMark(Range[] ranges, SecureMarks secureMarks)
         return newRanges.ToArray();
     }
 
-    internal static bool ContainsUnsafeRange(IEnumerable<Range>? ranges)
+    internal static Range[]? GetUnsafeRanges(Range[] ranges, SecureMarks safeMarks, SourceType[]? safeSources)
     {
-        if (ranges is null)
+        if (safeMarks == SecureMarks.None && safeSources is null)
         {
-            return false;
+            return ranges;
         }
 
-        foreach (var range in ranges)
+        int insecureCount = 0;
+        for (int x = 0; x < ranges.Length; x++)
+        {
+            var range = ranges[x];
+            if (!range.IsSecure(safeMarks, safeSources))
+            {
+                insecureCount++;
+            }
+        }
+
+        if (insecureCount == ranges.Length)
+        {
+            // This is made in order to avoid unnecessary allocations (most common situation)
+            return ranges;
+        }
+
```

Comment:
nit: could also cater to the easy one:

```suggestion

        if (insecureCount == 0)
        {
            return [];
        }
```

Reply:
Agree